File: CreationalDesignPattern/ObjectPool/java/ThreadPool.java
package CreationalDesignPattern.ObjectPool.java;

import java.util.ArrayList;

public class ThreadPool {
    private ArrayList<String> freeThreads;
    private ArrayList<String> inUseThreads;

    public ThreadPool(int size){
        this.freeThreads = new ArrayList<>(size);
        this.inUseThreads = new ArrayList<>(size);
        for(int i=0;i<size;i++){
            this.freeThreads.add("Thread-"+(i+1));
        }
    }

    public String getThread(){
        if(this.freeThreads.size() != 0){
            String threadToSend = this.freeThreads.get(0);
            this.inUseThreads.add(threadToSend);
            this.freeThreads.remove(threadToSend);
            return threadToSend;
        }
        return "No Free Threads";
    }

    public void releaseThread(String threadToRelease){
        if(this.inUseThreads.contains(threadToRelease)){
            this.inUseThreads.remove(threadToRelease);
            this.freeThreads.add(threadToRelease);
        }
    }
}

File: CreationalDesignPattern/ObjectPool/python/ThreadPool.py
class ThreadPool:
    def __init__(self, size):
        self.inUseThreads = []
        self.freeThreads = ["Thread-{}".format(i+1) for i in range(size)]
    
    def getThread(self):
        if len(self.freeThreads) != 0:
            threadToReturn = self.freeThreads[0]
            self.freeThreads.remove(threadToReturn)
            self.inUseThreads.append(threadToReturn)
            return threadToReturn
        return "No Free Threads"
    
    def releaseThread(self, threadToRelease):
        if threadToRelease in self.inUseThreads:
            self.inUseThreads.remove(threadToRelease)
            self.freeThreads.append(threadToRelease)


tpool = ThreadPool(2)
t1 = tpool.getThread()
print(t1)
t2 = tpool.getThread()
print(t2)
t3 = tpool.getThread()
print(t3)
tpool.releaseThread(t2)
t3 = tpool.getThread()
print(t3)
